### PR TITLE
boards: soc: riscv: telink_b9x: added flash size check

### PR DESCRIPTION
**Pull request by UR6LAL**
added flash size check

### CODE REVIEW

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 1]
```diff
@@ -239,3 +281,5 @@ void soc_b9x_restore(void)
 }
 
 SYS_INIT(soc_b9x_init, PRE_KERNEL_1, 0);
+
+SYS_INIT(soc_b9x_check_flash, APPLICATION, 0);
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: just check inside soc_b9x_init

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 2]
```diff
@@ -109,6 +116,41 @@ void soc_load_rf_parameters_deep_retention(void)
...
+		default:
+			break;
+		}
+		return status;
+}
+
+static int soc_b9x_check_flash(void)
+{
+	if (flash_cap_check() == false) {
+		k_fatal_halt(K_ERR_KERNEL_OOPS);
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: Add some readable message here.

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 3]
```diff
@@ -109,6 +116,41 @@ void soc_load_rf_parameters_deep_retention(void)
 }
 #endif
 
+bool flash_cap_check(void) {
+	bool status  = false;
+	uint32_t data = flash_read_mid();
+	uint8_t flash_cap = (uint8_t)(data >> FLASH_SIZE_OFFSET);
+	size_t bin_size = FIXED_PARTITION_SIZE(flash);
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: more clean is to use chosen
zephyr,flash

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 4]
```diff
@@ -19,6 +19,13 @@
 #include "b9x_bt_flash.h"
 #endif
 
+/* List of supported flash capacities */
+#define FLASH_1M_CAP 0x100000
+#define FLASH_2M_CAP 0x200000
+#define FLASH_4M_CAP 0x400000
+
+#define FLASH_SIZE_OFFSET 16
+
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: seems defines are prepared:
```
/**
 * @brief	flash capacity definition
 *			Call flash_read_mid function to get the size of flash capacity.
 *			Example is as follows:
 *			unsigned int mid = flash_read_mid();
 *			The value of (mid&0x00ff0000)>>16 reflects flash capacity.
 */
typedef enum {
    FLASH_SIZE_64K     = 0x10,
    FLASH_SIZE_128K    = 0x11,
    FLASH_SIZE_256K    = 0x12,
    FLASH_SIZE_512K    = 0x13,
    FLASH_SIZE_1M      = 0x14,
    FLASH_SIZE_2M      = 0x15,
    FLASH_SIZE_4M      = 0x16,
    FLASH_SIZE_8M      = 0x17,
	FLASH_SIZE_16M     = 0x18,
}flash_capacity_e;
```

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 5]
```diff
@@ -109,6 +116,41 @@ void soc_load_rf_parameters_deep_retention(void)
...
 #endif
 
+bool flash_cap_check(void) {
+	bool status  = false;
+	uint32_t data = flash_read_mid();
+	uint8_t flash_cap = (uint8_t)(data >> FLASH_SIZE_OFFSET);
+	size_t bin_size = FIXED_PARTITION_SIZE(flash);
+	switch(flash_cap) {
+		case FLASH_SIZE_1M:
+			if(FLASH_1M_CAP > bin_size) {
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: Don't use obvious macros. Just set number.

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 6]
```diff
@@ -109,6 +116,41 @@ void soc_load_rf_parameters_deep_retention(void)
 }
 #endif
 
+bool flash_cap_check(void) {
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: Rename to more clean way like flash_cap_suitable. In that way it's clean true/false meaning.

File: soc/riscv/riscv-privilege/telink_b9x/soc.c [hunk 7]
```diff
@@ -109,6 +116,41 @@ void soc_load_rf_parameters_deep_retention(void)
...
+	uint32_t data = flash_read_mid();
+	uint8_t flash_cap = (uint8_t)(data >> FLASH_SIZE_OFFSET);
+	size_t bin_size = FIXED_PARTITION_SIZE(flash);
+	switch(flash_cap) {
+		case FLASH_SIZE_1M:
+			if(FLASH_1M_CAP > bin_size) {
+				status = true;
+			}
+			break;
+		case FLASH_SIZE_2M:
```
Comment (andriy-bilynskyy) [inline soc/riscv/riscv-privilege/telink_b9x/soc.c]: set from 1 till 16

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 3 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: Thank you, changed to DT_REG_SIZE(DT_CHOSEN(zephyr_flash)

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 2 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: Added the message with the crash reason

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 1 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: I can't use it in PRE_KERNEL (we don't have logs at this stage). Moved to earlier stage then APPLICATION - POST_KERNEL

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 4 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: Yes, I use a similar method to get flash_capacity

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 5 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: Fixed

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 6 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: fixed

File: soc/riscv/riscv-privilege/telink_b9x/soc.c (hunk 7 above)
Comment (UR6LAL) [reply soc/riscv/riscv-privilege/telink_b9x/soc.c]: added

**Review by andriy-bilynskyy — APPROVED** [review]

**Pull request merged**